### PR TITLE
FORMS-17028: adding recaptcha component to UI

### PR DESCRIPTION
**Pull request by lovelymandal16**
This PR creates a recaptcha component in UE under adaptive forms option.
The contains unit test for enterprise recaptcha.

Fix #FORMS-17028

Test URLs:
- Before: https://main--aem-boilerplate-forms--adobe-rnd.hlx.live/
- After: https://recaptchaenterprise--aem-boilerplate-forms--adobe-rnd.hlx.page/

### CODE REVIEW

File: blocks/form/form.js [hunk 1]
```diff
@@ -394,6 +395,17 @@ async function createFormForAuthoring(formDef) {
   return form;
 }
 
+function getSitePageName(path) {
```
Comment (nit23uec) [inline blocks/form/form.js]: this would not work for doc based forms. pls verify the flow for doc based forms also.

File: blocks/form/form.js [hunk 2]
```diff
@@ -406,8 +418,9 @@ export async function createForm(formDef, data) {
 
   let captcha;
   if (captchaField) {
-    const siteKey = captchaField?.properties?.['fd:captcha']?.config?.siteKey || captchaField?.value;
-    captcha = new GoogleReCaptcha(siteKey, captchaField.id);
+    const config = captchaField?.properties?.['fd:captcha']?.config;
```
Comment (nit23uec) [inline blocks/form/form.js]: this wont work for doc based forms.

File: blocks/form/form.js (hunk 1 above)
Comment (nit23uec) [inline blocks/form/form.js]: better to move it to util.js

File: models/form/_recaptcha.json [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+            "id": "captcha",
+            "fields": [
+                {
+                    "component": "container",
+                    "name": "basic",
+                    "label": "Basic",
+                    "collapsible": false,
+                    "fields": [
+                        {
+                            "...": "./_common.json#/basicfields"
```
Comment (nit23uec) [inline models/form/_recaptcha.json]: do all the common props make sense for captcha?

File: scripts/form-editor-support.js [hunk 4]
```diff
@@ -96,6 +96,32 @@ function annotateFormFragment(fragmentFieldWrapper, fragmentDefinition) {
...
+  captchaFieldWrapper.appendChild(titleEl);
+}
+
+function annotateRecaptcha(captchaFieldWrapper, recaptchaDefinition) {
+  captchaFieldWrapper.classList.toggle('captcha-wrapper', true);
+  if (document.documentElement.classList.contains('adobe-ue-edit')) {
+    const newFieldWrapper = captchaFieldWrapper.cloneNode(true);
+    newFieldWrapper.setAttribute('data-aue-type', 'component');
+    newFieldWrapper.setAttribute('data-aue-resource', `urn:aemconnection:${recaptchaDefinition.properties['fd:path']}`);
+    newFieldWrapper.setAttribute('data-aue-model', 'recaptcha-v1');
```
Comment (nit23uec) [inline scripts/form-editor-support.js]: model should be `captcha` only.

File: scripts/form-editor-support.js [hunk 5]
```diff
@@ -96,6 +96,32 @@ function annotateFormFragment(fragmentFieldWrapper, fragmentDefinition) {
...
+  titleEl.classList.add('captcha-title');
+  titleEl.textContent = fragmentDefinition.label?.value || fragmentDefinition.name;
+  titleEl.id = 'recaptcha-title';
+  captchaFieldWrapper.appendChild(titleEl);
+}
+
+function annotateRecaptcha(captchaFieldWrapper, recaptchaDefinition) {
+  captchaFieldWrapper.classList.toggle('captcha-wrapper', true);
+  if (document.documentElement.classList.contains('adobe-ue-edit')) {
+    const newFieldWrapper = captchaFieldWrapper.cloneNode(true);
```
Comment (nit23uec) [inline scripts/form-editor-support.js]: why do need to clone the node for captcha?
can;t we just add edit-mode class to the existing dom element?

File: scripts/form-editor-support.js [hunk 6]
```diff
@@ -145,6 +171,8 @@ function annotateItems(items, formDefinition, formFieldMap) {
                 handleWizardNavigation(fieldWrapper.parentElement, fieldWrapper);
               }
             }
+          } else if (fd.fieldType === 'captcha') {
+            annotateRecaptcha(fieldWrapper, fd);
```
Comment (nit23uec) [inline scripts/form-editor-support.js]: I dont think we need this code.

**Review by nit23uec — CHANGES_REQUESTED** [review]
comments inline

File: blocks/form/form.js (hunk 1 above)
Comment (lovelymandal16) [reply blocks/form/form.js]: made some changes to fetch config for doc-based form. The working demo of recaptcha in doc-based form is posted on the scrum channel. Please have a look, and let me know your thoughts

File: blocks/form/form.js (hunk 2 above)
Comment (lovelymandal16) [reply blocks/form/form.js]: i Have made changes to fetch config. Debugged with doc-based form, present code seem to be working. Here is the google doc link : https://drive.google.com/drive/u/0/folders/1ogYjvnxzQn8cF5MH-7FYRIvpWQZNK-kB

File: blocks/form/form.js (hunk 1 above)
Comment (lovelymandal16) [reply blocks/form/form.js]: Moved it to utils.js

File: scripts/form-editor-support.js (hunk 4 above)
Comment (lovelymandal16) [reply scripts/form-editor-support.js]: Reverted changes from form-editor-support.js

File: scripts/form-editor-support.js (hunk 5 above)
Comment (lovelymandal16) [reply scripts/form-editor-support.js]: Reverted all changes from form-editor-support.js

File: scripts/form-editor-support.js (hunk 6 above)
Comment (lovelymandal16) [reply scripts/form-editor-support.js]: Reverted all changes from form-editor-support.js

**Comment by nit23uec** [comment]
can you pls run `npm run build:json` and then commit the changes?

File: models/form/_recaptcha.json [hunk 7]
```diff
@@ -0,0 +1,43 @@
+{
+    "definitions": [
+        {
+            "title": "CAPTCHA",
```
Comment (nit23uec) [inline models/form/_recaptcha.json]: I think we need to have the title that's visible in the Content Tree as: `Captcha (Invisible)`

File: models/form/_recaptcha.json [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+            "id": "captcha",
+            "plugins": {
+              "xwalk": {
+                "page": {
+                  "resourceType": "core/fd/components/form/recaptcha/v1/recaptcha",
+                  "template": {
+                    "jcr:title": "CAPTCHA",
+                    "fieldType": "captcha",
+                    "enabled": true,
```
Comment (nit23uec) [inline models/form/_recaptcha.json]: we can skip enabled and visible props here.

File: models/form/_recaptcha.json [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+                  }
+                }
+              }
+            }
+        }
+    ],
+    "models": [
+        {
+            "id": "captcha",
+            "fields": [
```
Comment (nit23uec) [inline models/form/_recaptcha.json:22]: we need not encapsulate the props in a container as there is only a single prop.

File: models/form/_recaptcha.json (hunk 3 above)
Comment (lovelymandal16) [reply models/form/_recaptcha.json]: remote all prompts, only kept name

File: models/form/_recaptcha.json (hunk 9 above)
Comment (lovelymandal16) [reply models/form/_recaptcha.json:22]: I tried that, but it caused issue in the rendering, as it expects an iterable. Here is a screenshot of the error. So I kept it as is.
image

File: models/form/_recaptcha.json (hunk 8 above)
Comment (lovelymandal16) [reply models/form/_recaptcha.json]: modified

File: models/form/_recaptcha.json (hunk 9 above)
Comment (nit23uec) [reply models/form/_recaptcha.json:22]: Pls look at text or image components

Comment (lovelymandal16) [reply models/form/_recaptcha.json:22]: _text.json and _image.json both have arrays (similar to the implementation for _recaptcha.son done here). The pre-commit hook anyway creates component-definition.json and does not have the container.

Comment (nit23uec) [reply models/form/_recaptcha.json:22]: it should not have the basic container. just the array of fields directly similar to what has been done for text and image.

Comment (lovelymandal16) [reply models/form/_recaptcha.json:22]: removed basic container from model/properties

**Review by nit23uec — APPROVED** [review]

**Pull request merged**